File: JavaScript/dfa.js
class State {
    constructor(is_start, is_final, name) {
	this.is_start = is_start;
	this.is_final = is_final;
	this.name = name;
    }
}

class DFA {
    constructor(alphabet) {
	this.sigma = alphabet;
	this.transitions = {};
	this.states = {};
	this.rejected = 0;

	this.curr_state = null;
    }

    add_state(is_start, is_final, name) {
	if(!this.states[name]) {
	    this.states[name] = new State(is_start, is_final, name);
	} else {
	    alert("State already exists!");
	}
    }

    add_transition(curr_state, input_symbol, next_state) {
	if(!this.sigma.includes(input_symbol)) {
	    alert("Input symbol not part of the alphabet");
	}
	if(!curr_state in this.states) {
	    alert("Current set is not part of the existing set of states");
	}
	if(!next_state in this.states) {
	    alert("Next state is not part of the existing set of states");
	}

	if(!this.transitions[input_symbol]) {
	    this.transitions[input_symbol] = [];
	}

	this.transitions[input_symbol].push([curr_state, input_symbol, next_state]);
    }

    get_start_state() {
	for(let state in this.states) {
	    if(this.states[state].is_start == 1) {
		return this.states[state];
	    }
	}
    }

    get_final_states() {
	let final_states = [];
	for(let state in this.states) {
	    if(this.states[state].is_final == 1) {
		final_states.push(this.states[state].name);
	    }
	}
	return final_states;
    }

    apply_transition(partial_transition) {
	let symbol = partial_transition[1];
	let curr_state = partial_transition[0];

	let applicable_transitions = [];
	for(let i = 0; i < this.transitions[symbol].length; i++) {
	    if(this.transitions[symbol][i][0] === curr_state) {
		applicable_transitions.push(this.transitions[symbol][i]);
	    }
	}

	if(applicable_transitions.length != 0) {
	    this.curr_state = this.states[applicable_transitions[0][2]];
	} else {
	    this.curr_state = this.states[curr_state];
	    this.rejected = 1;
	}
    }

    evaluate_string(string) {
	let set = new Set(string);
	for(let i=0;i<[...set].length;i++) {
	    if(!this.sigma.includes([...set][i])) {
		alert("Symbol not part of the alphabet");
		return null;
	    }
	}

	this.rejected = 0;
	let start_state = this.get_start_state();
	let final_states = this.get_final_states();

	this.curr_state = start_state;
	let string_arr = [...string];
	for(let i=0;i<string_arr.length;i++) {
	    if(this.rejected != 1) {
		this.apply_transition([this.curr_state.name, string_arr[i]]);
	    }
	    else {
		alert("Rejected");
		return null;
	    }
	}

	if(final_states.includes(this.curr_state.name)) {
	    alert("Accepted!");
	} else {
	    alert("Rejected");
	}
    }
}

let dfa = new DFA(['a','b']);

dfa.add_state(1, 1, 'q0');
dfa.add_state(0, 1, 'q1');
dfa.add_state(0, 1, 'q2');
dfa.add_state(0, 1, 'q3');
dfa.add_state(0, 0, 'q4');


dfa.add_transition('q0', 'a', 'q1');
dfa.add_transition('q0', 'b', 'q0');
dfa.add_transition('q1', 'a', 'q2');
dfa.add_transition('q1', 'b', 'q1');
dfa.add_transition('q2', 'a', 'q3');
dfa.add_transition('q2', 'b', 'q2');
dfa.add_transition('q3', 'a', 'q4');
dfa.add_transition('q3', 'b', 'q3');
dfa.add_transition('q4', 'a', 'q4');
dfa.add_transition('q4', 'b', 'q4');

dfa.evaluate_string("aaab");
dfa.evaluate_string("aaaab");

File: Python/dfa.py
class State:
    def __init__(self, is_start, is_final, name):
        """
        is_start: Whether the state is a start state or not (0 or 1)
        is_final: Whether the state is a final state or not (0 or 1)
        name : Name of the state, expected to be unique
        """
        self.is_final = is_final
        self.is_start = is_start
        self.name = name
    
class DFA:
    def __init__(self, alphabet):
        """
        sigma: Set of symbols accepted by the language of the DFA
        transitions: Each transition is of the form (current state, input symbol, next state) 
                     A dictionary with input symbols as keys is maintained for transitions
                     Each key will have a list of tuple(s)
        states: Set of states of the DFA
                A dictionary with state names as keys and values as an object of the class State
        rejected: Flag to indicated whether a state ends up "dead" or there is not defined 
                  transisiton for the current input symbol
        curr_state : Keeps track of the state the DFA is currently in after each transistion 
        __num_start_states : Number of states marked as start state in the DFA.
        """
        self.sigma = alphabet        
        self.transitions = {}
        self.states = {}
        self.rejected = 0
        
        self.curr_state = None
        self.__num_start_states = 0
    
    def add_state(self, is_start, is_final, name):
        if name not in self.states:
            self.states[name] = State(is_start, is_final, name)

            if is_start:
                self.__num_start_states += 1
            
            if self.__num_start_states != 1:
                #This is done in this fashion from a user-convinience point of view.
                # Say, a user has mistakenly set a previous node as start when actually the current node is
                # the start node. In that case, if the current node is rejected directly, that means they can't add
                # the actual state node until they resolve conflict. This way, they have additional 
                # flexibility on this matter.
                print("More than one start state detected but node added")
        else:
            print("State already exists!")
    
    def add_transition(self, curr_state, input_symbol, next_state):
        if input_symbol not in self.sigma:
            print("Input symbol not part of alphabet")
            return None
        
        if curr_state not in self.states:
            print("Current state is not part of the existing set of states")
            return None
        if next_state not in self.states:
            print("Next state is not part of the existing set of states")
            return None
        
        if input_symbol not in self.transitions:
            self.transitions[input_symbol] = []
        
        self.transitions[input_symbol].append((curr_state, input_symbol, next_state))
    
    def get_start_state(self):
        for state in self.states:
            if self.states[state].is_start:
                return self.states[state]
            
    def get_final_states(self):
        final = []
        for state in self.states:
            if self.states[state].is_final:
                final.append(self.states[state].name)
                
        return final
    
    def apply_transition(self, partial_transition):
        symbol = partial_transition[1]
        curr_state = partial_transition[0]
        
        applicable_transitions = [i for i in self.transitions[symbol] if curr_state == i[0]]

        if len(applicable_transitions):
            self.curr_state = self.states[applicable_transitions[0][2]]
        else:
            self.curr_state = self.states[curr_state]
            self.rejected = 1
    
    def evaluate_string(self, string):

        if self.__num_start_states != 1:
            print("More than one start state detected. Exiting")
            return

        for symbol in list(set(string)):
            if symbol not in self.sigma:
                print("'"+symbol+"' is not part of the alphabet")
                return
        
        self.rejected = 0
        start_state = self.get_start_state()
        final_states = self.get_final_states()
        
        self.curr_state = start_state
        
        for symbol in list(string):
            if not self.rejected:
                self.apply_transition((self.curr_state.name, symbol))
            else:
                print("Rejected")
                return
        
        if self.curr_state.name in final_states:
            print("Accepted!")
        else:
            print("Rejected")


dfa = DFA(['a','b'])

#add states to the DFA
dfa.add_state(1, 1, 'q0')
dfa.add_state(0, 1, 'q1')
dfa.add_state(0, 1, 'q2')
dfa.add_state(0, 1, 'q3')
dfa.add_state(0, 0, 'q4')

#add transitions to the DFA 
dfa.add_transition('q0', 'a', 'q1')
dfa.add_transition('q0', 'b', 'q0')
dfa.add_transition('q1', 'a', 'q2')
dfa.add_transition('q1', 'b', 'q1')
dfa.add_transition('q2', 'a', 'q3')
dfa.add_transition('q2', 'b', 'q2')
dfa.add_transition('q3', 'a', 'q4')
dfa.add_transition('q3', 'b', 'q3')
dfa.add_transition('q4', 'a', 'q4')
dfa.add_transition('q4', 'b', 'q4')

dfa.evaluate_string("aaab")  #Accepted!
dfa.evaluate_string("aaaab") #Rejected
dfa.evaluate_string("")	     #Accepted!
dfa.evaluate_string("bbbb")  #Accepted!
